clang/lib/3C/CastPlacement.cpp

```diff
@@ -18,107 +18,244 @@
-              std::string CastString = getCastString(ArgumentC, ParameterC);
-              surroundByCast(CastString, A);
-              break;
-            }
+        CVarSet ArgumentConstraints = CR.getExprConstraintVars(ArgExpr);
+        ConstraintVariable *ParamInt = FV->getInternalParam(PIdx);
+        ConstraintVariable *ParamExt = FV->getExternalParam(PIdx);
+        for (auto *ArgC : ArgumentConstraints) {
+          if (needCasting(ArgC, ArgC, ParamInt, ParamExt) != NO_CAST) {
+            surroundByCast(ArgC, ArgC, ParamInt, ParamExt, A);
+            break;
           }
         }
-        PIdx++;
+      }
+      PIdx++;
+    }
+
+    // Cast on return
+    if (Info.hasPersistentConstraints(CE, Context)) {
+      CVarSet ArgumentConstraints = CR.getExprConstraintVars(CE);
+      ConstraintVariable *RetInt = FV->getInternalReturn();
+      ConstraintVariable *RetExt = FV->getExternalReturn();
+      for (auto *ArgC : ArgumentConstraints) {
+        // Order of ParameterC and ArgumentC is reversed from when inserting
+        // parameter casts because assignment now goes from returned to its
+        // local use.
+        if (!Locator.exprHasCast(CE) &&
+            needCasting(RetInt, RetExt, ArgC, ArgC) != NO_CAST) {
+          surroundByCast(RetInt, RetExt, ArgC, ArgC, CE);
+          break;
+        }
       }
     }
   }
   return true;
 }
 
 // Check whether an explicit casting is needed when the pointer represented
-// by src variable is assigned to dst.
-bool CastPlacementVisitor::needCasting(ConstraintVariable *Src,
-                                       ConstraintVariable *Dst) {
+// by src variable is assigned to dst. The return value specifies exactly the
+// type of cast required.
+CastPlacementVisitor::CastNeeded
+CastPlacementVisitor::needCasting(const ConstraintVariable *SrcInt,
+                                  const ConstraintVariable *SrcExt,
+                                  const ConstraintVariable *DstInt,
+                                  const ConstraintVariable *DstExt) {
+
+  if (isa<PVConstraint>(DstExt) && cast<PVConstraint>(DstExt)->isVoidPtr())
+    return CastNeeded::NO_CAST;
+
+  if (isa<FVConstraint>(SrcExt) &&
+      SrcExt->solutionEqualTo(Info.getConstraints(), DstExt))
+    return CastNeeded::NO_CAST;
+
   const auto &E = Info.getConstraints().getVariables();
-  // Check if the src is a checked type.
-  if (Src->isChecked(E)) {
-    // If Dst has an itype, Src must have exactly the same checked type. If this
-    // is not the case, we must insert a case.
-    if (Dst->hasItype())
-      return !Dst->solutionEqualTo(Info.getConstraints(), Src);
-
-    // Is Dst Wild?
-    // TODO: The Dinfo == WILD comparison seems to be the cause of a cast
-    //       insertion bug. Can it be removed?
-    if (!Dst->isChecked(E))
-      return true;
+  bool SIChecked = SrcInt->isFullyChecked(E) && !SrcInt->srcHasItype();
+  bool SEChecked = SrcExt->isFullyChecked(E) || SrcExt->srcHasItype();
```

Comment:
I'm confused about the logic here. What is `isFullyChecked` for itypes? The first line seems to imply you could be fully checked but have an itype, but the second implies you could not be fully checked, but have one. This seems contradictory? (Can't immediately tell what `E` is, which could help.)

Reply:
In general, I've lost context as to why we need to know the internal/external for a source and internal/external for a dest. Why wouldn't it always just be internal -> external or vice versa (and you can reverse the polarity as needed) ?